provider/oracle/environ.go

```diff
@@ -0,0 +1,731 @@
+		if err != nil {
+			return nil, errors.Trace(err)
+		}
+		ret = append(ret, oInst)
+		return ret, nil
+	}
+
+	resp, err := o.client.AllInstances(nil)
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+
+	if len(resp.Result) == 0 {
+		return nil, environs.ErrNoInstances
+	}
+
+	for _, val := range resp.Result {
+		for _, id := range ids {
+			if val.Name == string(id) {
+				oInst, err := newInstance(val, o)
+				if err != nil {
+					return nil, errors.Trace(err)
+				}
+				ret = append(ret, oInst)
+				break
+			}
+		}
+	}
+	if len(ret) < len(ids) {
+		return ret, environs.ErrPartialInstances
+	}
+	return ret, nil
+}
+
+func (o *oracleEnviron) getOracleInstancesAsMap(ids ...instance.Id) (map[string]*oracleInstance, error) {
+	instances, err := o.getOracleInstances(ids...)
+	if err != nil {
+		return map[string]*oracleInstance{}, errors.Trace(err)
+	}
+	ret := map[string]*oracleInstance{}
+	for _, val := range instances {
+		ret[string(val.Id())] = val
+	}
+	return ret, nil
+}
+
+// AllInstances is part of the InstanceBroker interface.
+func (o *oracleEnviron) AllInstances() ([]instance.Instance, error) {
+	tagFilter := tagValue{tags.JujuModel, o.Config().UUID()}
+	instances, err := o.allInstances(tagFilter)
+	if err != nil {
+		return nil, err
+	}
+
+	ret := make([]instance.Instance, len(instances))
+	for i, val := range instances {
+		ret[i] = val
+	}
+	return ret, nil
+}
+
+func (o *oracleEnviron) allInstances(tagFilter tagValue) ([]*oracleInstance, error) {
+	filter := []oci.Filter{
+		oci.Filter{
+			Arg:   "tags",
+			Value: tagFilter.String(),
+		},
+	}
+	logger.Infof("Looking for instances with tags: %v", filter)
+	resp, err := o.client.AllInstances(filter)
+	if err != nil {
+		return nil, err
+	}
+
+	n := len(resp.Result)
+	instances := make([]*oracleInstance, 0, n)
+	for _, val := range resp.Result {
+		oracleInstance, err := newInstance(val, o)
+		if err != nil {
+			return nil, errors.Trace(err)
+		}
+		instances = append(instances, oracleInstance)
+	}
+
+	return instances, nil
+}
+
+// MaintainInstance is part of the InstanceBroker interface.
+func (o *oracleEnviron) MaintainInstance(args environs.StartInstanceParams) error {
+	return nil
+}
+
+// Config is part of the Environ interface.
+func (o *oracleEnviron) Config() *config.Config {
+	o.mutex.Lock()
+	defer o.mutex.Unlock()
+	return o.cfg
+}
+
+// ConstraintsValidator is part of the environs.Environ interface.
+func (o *oracleEnviron) ConstraintsValidator() (constraints.Validator, error) {
+	// list of unsupported oracle provider constraints
+	unsupportedConstraints := []string{
+		constraints.Container,
+		constraints.CpuPower,
+		constraints.RootDisk,
+		constraints.VirtType,
+		constraints.Spaces,
+	}
+
+	// we choose to use the default validator implementation
+	validator := constraints.NewValidator()
+	// we must feed the validator that the oracle cloud
+	// provider does not support these constraints
+	validator.RegisterUnsupported(unsupportedConstraints)
```

Comment:
You must register vocabulary for constraints.Arch, to support bootstrapping with "--build-agent". Presumably Oracle Cloud will only ever support x86-64 (maybe i386?), so you can just hard-code this:
`validator.RegisterVocabulary(constraints.Arch, []string{arch.I386, arch.AMD64})`